test/s3-test-srv/main.go

```diff
@@ -2,28 +2,89 @@ package main
 
 import (
 	"flag"
+	"fmt"
 	"io"
 	"log"
 	"net/http"
+	"sync"
 
 	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/crl/crl_x509"
+	"github.com/letsencrypt/boulder/revocation"
 )
 
+type s3TestSrv struct {
+	sync.Mutex
+	allSerials map[string]revocation.Reason
+}
+
+func (srv *s3TestSrv) handleUpload(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(500)
+		w.Write([]byte("failed to read request body"))
+		return
+	}
+
+	crl, err := crl_x509.ParseRevocationList(body)
+	if err != nil {
+		w.WriteHeader(500)
+		w.Write([]byte(fmt.Sprintf("failed to parse body: %s", err)))
+		return
+	}
+
+	srv.Lock()
+	defer srv.Unlock()
+	for _, rc := range crl.RevokedCertificates {
+		reason := 0
+		if rc.ReasonCode != nil {
+			reason = *rc.ReasonCode
+		}
+		srv.allSerials[core.SerialToString(rc.SerialNumber)] = revocation.Reason(reason)
+	}
+
+	w.WriteHeader(200)
+	w.Write([]byte("{}"))
+}
+
+func (srv *s3TestSrv) handleClear(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		w.WriteHeader(405)
+		return
+	}
+
+	srv.Lock()
+	defer srv.Unlock()
+	srv.allSerials = make(map[string]revocation.Reason)
+}
+
+func (srv *s3TestSrv) handleQuery(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		w.WriteHeader(405)
+		return
+	}
+
+	serial := r.URL.Query().Get("serial")
+	reason, ok := srv.allSerials[serial]
```

Comment:
Nit, if a query parameter is missing, you should respond with a `400`

```suggestion
	serial := r.URL.Query().Get("serial")
	if serial == "" {
		w.WriteHeader(400)
		w.Write([]byte("missing 'serial' query parameter"))
		return
	}
	reason, ok := srv.allSerials[serial]
```

Reply:
It will -- looking up the empty string in the map will fail, resulting in a 404.

Reply:
> It will -- looking up the empty string in the map will fail, resulting in a 404.

It will `404 Not Found` not `400 Bad Request` as I'm suggesting.

Reply:
More specifically, per the [docs](https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html):



MissingRequiredParameter | The SelectRequest entity is missing a required parameter. Check the service documentation and try again. | 400
-- | -- | --